geomesa-convert/geomesa-convert-json/src/main/scala/org/locationtech/geomesa/convert/json/JsonConverterFactory.scala

```diff
@@ -9,138 +9,200 @@
   JsonConverterFactory.TypeToProcess, JsonConfigConvert, JsonFieldConvert, BasicOptionsConvert) {
 
+  import scala.collection.JavaConverters._
+
+  override def infer(
+      is: InputStream,
+      sft: Option[SimpleFeatureType],
+      path: Option[String]): Option[(SimpleFeatureType, Config)] =
+    infer(is, sft, path.map(EvaluationContext.inputFileParam).getOrElse(Map.empty)).toOption
+
+  /**
+   * Infer a configuration and simple feature type from an input stream, if possible
+   *
+   * Available hints:
+   *  - `featurePath` - json path expression pointing to the feature element
+   *
+   * @param is input
+   * @param sft simple feature type, if known ahead of time
+   * @param hints implementation specific hints about the input
+   * @return
+   */
   override def infer(
       is: InputStream,
       sft: Option[SimpleFeatureType],
-      path: Option[String]): Option[(SimpleFeatureType, Config)] = {
-    try {
-      val reader = new JsonReader(new InputStreamReader(is, StandardCharsets.UTF_8))
-      reader.setLenient(true)
+      hints: Map[String, AnyRef]): Try[(SimpleFeatureType, Config)] = {
 
-      val elements = {
+    val tryElements = Try {
+      WithClose(new JsonReader(new InputStreamReader(is, StandardCharsets.UTF_8))) { reader =>
+        reader.setLenient(true)
         val iter: Iterator[JsonElement] = new Iterator[JsonElement] {
-          private val parser = new JsonParser
           override def hasNext: Boolean = reader.peek() != JsonToken.END_DOCUMENT
-          override def next(): JsonElement = parser.parse(reader)
+          override def next(): JsonElement = JsonParser.parseReader(reader)
         }
-        iter.take(AbstractConverterFactory.inferSampleSize).toSeq
+        iter.take(AbstractConverterFactory.inferSampleSize).toList
       }
-
-      val geojson = elements.collect {
-        case el if JsonConverter.isFeature(el) => JsonConverter.parseFeature(el)
-        case el if JsonConverter.isFeatureCollection(el) => JsonConverter.parseFeatureCollection(el)
+    }
+    tryElements.flatMap { elements =>
+      lazy val featurePath = hints.get(JsonConverterFactory.FeaturePathKey).map(_.toString)
+      if (elements.isEmpty) {
+        Failure(new RuntimeException("Could not parse the input as JSON"))
+      } else if (JsonConverter.isFeature(elements.head)) {
+        Try(elements.map(JsonConverter.parseFeature))
+            .flatMap(inferGeoJson(_, None, sft))
+      } else if (JsonConverter.isFeatureCollection(elements.head)) {
+        Try(elements.flatMap(JsonConverter.parseFeatureCollection))
+            .flatMap(inferGeoJson(_, Some("$.features[*]"), sft))
+      } else if (elements.head.isJsonObject) {
+        inferJson(elements, featurePath, sft)
+      } else if (elements.head.isJsonArray) {
+        inferJson(elements, featurePath.orElse(Some("$.[*]")), sft)
+      } else {
+        Failure(new RuntimeException("Could not parse the input as a JSON object or array"))
```

Comment:
We might want to implement a unit test for this failure case, if we don't already have one.